packages/vite/src/node/optimizer/index.ts

```diff
@@ -62,6 +63,11 @@ export interface DepOptimizationOptions {
    * cannot be globs).
    */
   exclude?: string[]
+  /**
+   * Force ESM interop when importing for these dependencies. Some legacy
+   * packages advertise themselves as ESM but use `require` internally
+   */
+  needsInterop?: string[]
```

Comment:
Perhaps to prevent a breaking change at the meantime, we could have this option opt-in? E.g. by default Vite will still use the old method, but if the user set `needsInterop: []` it will use the new optimization.

Also agree that `acceptMixedSyntax` is a bit limiting.

PS: We'd also need this documented in the docs.

Reply:
[19e7168](https://github.com/vitejs/vite/pull/7568/commits/19e7168cbe1c53d9f20dd52d43e3676f99b76a55) avoids the breaking change. Mixed ESM and CJS would still work, and the warning and hint to add them to `optimizeDeps.needsInterop` have been improved.
The only diff with them is that cold start is a bit slower because of an extra page reload if they aren't added to `optimizeDeps.needsInterop`.
With this change, I think the tradeoff makes more sense... we are making things faster for everybody that doesn't use the legacy package, but they still work (and adding them to the config makes things as fast as a normal package)

Reply:
Wait before checking again, I need to fix and simplify some things

Reply:
Ok, now it is ready. I think the code is easier to follow now. Let me know what you think @bluwy

Reply:
Sweet this sounds like a neat solution to me. Now it's fast by default.